Update Unit Test Best Practices (#619)
department-of-veterans-affairs/veteran-facing-services-tools: packages/documentation/src/pages/getting-started/common-tasks/new-unit-test.mdx

title: Writing a unit test
tags: Mocha, Chai, Enzyme, schemaform-utils
---

> 

Reviewer: I know all of these empty lines are probably unnecessary, but they do make it easier to read/edit the .mdx file. I'd recommend leaving them in.

Author: - [x] Add empty lines where it may serve to improve readability

All new code that is added to `vets-website` should be unit tested and unit tests should cover at least 75% of code paths. Write unit tests as you build to make sure your form (or other component) is behaving as you expect and to help guard against future bugs.
For example, you might create a unit test file for each page in a form and then test the following scenarios:
- The correct number of inputs show up when you render the page.
- The correct number of fields display validation errors if you submit without entering any information.
- Any conditional logic on the page displays under the correct conditions.
## Unit test overview
- The `vets-website` repo uses:
  - [Mocha](https://mochajs.org/) for running unit tests
  - [Chai](http://chaijs.com/) for test assertions
  - [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) and [Enzyme](https://airbnb.io/enzyme/) for mounting and inspecting React components

>     - We encourage the use of React Testing Library in place of Enzyme. [See the migration docs.] (https://testing-library.com/docs/react-testing-library/migrate-from-enzyme)

Reviewer: Nice addition here.

---

# Writing a unit test
All new code that is added to `vets-website` should be unit tested and unit tests should cover at least 75% of code paths. Write unit tests as you build to make sure your form (or other component) is behaving as you expect and to help guard against future bugs.
For example, you might create a unit test file for each page in a form and then test the following scenarios:
- The correct number of inputs show up when you render the page.
- The correct number of fields display validation errors if you submit without entering any information.
- Any conditional logic on the page displays under the correct conditions.

## Unit test overview
- The `vets-website` repo uses:
  - [Mocha](https://mochajs.org/) for running unit tests
  - [Chai](http://chaijs.com/) for test assertions

>   - [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) and [Enzyme](https://airbnb.io/enzyme/) for mounting and inspecting React components

Reviewer 2: If you refer to it as "React Testing Library (RTL)" here, I think it should be fine to refer to it as RTL later in the doc.

>     - We encourage the use of React Testing Library in place of Enzyme. [See the migration docs.] (https://testing-library.com/docs/react-testing-library/migrate-from-enzyme)

Reviewer 3: ```suggestion
    - We encourage the use of React Testing Library in place of Enzyme. [See the migration docs](https://testing-library.com/docs/react-testing-library/migrate-from-enzyme).
```

Author: - [x] Move period

      - 📂 components
        - MyComponent.unit.spec.jsx
```
Mocha runs any *file_name*.unit.spec.js file located in the `/src` folder. This file is usually located in a test directory close to the code being tested.
[Run the test](/getting-started/common-tasks/test) locally through npm script commands, during the Jenkins build (Unit), and after merging to master.
   
## Unit test conventions
Use the following conventions when writing a unit test.
```js
import React from 'react';
import { expect } from 'chai';
import { render, screen } from '@testing-library/react';
import MyComponent from '../../components/MyComponent';
describe('my-application', () => {
  describe('MyComponent', () => {
    it('renders privacy act disclosure when "show" is true', () => {
        render(<MyComponent show />);
        expect(screen.getByRole('heading')).to.have.text('Privacy Act Disclosure');
    });
  });
});
```

> - Use `describe` to organize tests by application and feature. Two levels is adequate depending on the size of the application being tested. Try to avoid nesting deeper than two levels.

Reviewer: ```suggestion
- Use `describe` to organize tests by application and feature. Two levels should be adequate depending on the size of the application being tested. Try to avoid nesting deeper than two levels.
```

Author: [x] Use "should be" in place of "is" to provide more flexibility

  - **Not recommended**:
    - `it('shortens the address when the user has a long address')`
    - `it('shows an error when the user is not logged in')`

>   - Use RTL's [render](https://testing-library.com/docs/react-testing-library/api/#render) when testing components.

Reviewer 3: ```suggestion
  - Use [React Testing Library's `render` function](https://testing-library.com/docs/react-testing-library/api/#render) when testing components.
```

Author: - [x] Spell out RTL

  - If you're using Enzyme, use [shallow](https://airbnb.io/enzyme/docs/api/shallow.html) instead of [mount](https://airbnb.io/enzyme/docs/api/mount.html) when possible to test components.

>     - Always `unmount` components at the end of the test.

Reviewer 2: I would put these under a subheading for this section about `Rendering components` along with examples of each. It'll be a cleaner separation from the previous points and give you more space for code blocks (less cluttered than if it were still in the list).

Author: - [x] Create new subsection with examples

    </div>
  );
};
export default SimpleLoginForm;
```
```js
// SimpleLoginForm.unit.spec.jsx
import React from 'react';
import { expect } from 'chai';
import sinon from 'sinon';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SimpleLoginForm from '../../components/SimpleLoginForm';
describe('my-application', () => {
  describe('SimpleLoginForm', () => {
    it('calls onSubmit with the username and password when submit is clicked', () => {
        const handleSubmit = sinon.spy();
        const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
        const user = { username: 'user123', password: 'password123' };
        userEvent.type(getByLabelText(/username/i), user.username);
        userEvent.type(getByLabelText(/password/i), user.password);
        userEvent.click(getByText(/submit/i));
        expect(handleSubmit.callCount).to.equal(1); // alternatively `expect(handleSubmit.calledOnce).to.be.true()` works as well
        expect(handleSubmit.calledWith(user)).to.be.true(); // for more explicit testing we can use `calledWithExactly` in place of `calledWith`
    });
  });
});
```

> We have successfully performed a "Happy Path" test for this component. Let's breakdown this test:

Reviewer: ```suggestion
We have successfully performed a "Happy Path" test for this component. Let's break down this test:
```

Author: Done

Author: - [x] Split "breakdown" into two words

```js
// Setup
...
const handleSubmit = sinon.spy();
const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
const user = { username: 'user123', password: 'password123' };
...
```

> The setup for this test is self-explanatory. We create a mocked submit function and pass that to our form component, and use the [render](https://testing-library.com/docs/react-testing-library/api/#render) function to produce actual DOM nodes. Then we define user data for reuse later on.

Reviewer: ```suggestion
The setup for this test is self-explanatory. We create a mocked submit function and pass that to our form component, then use the [render](https://testing-library.com/docs/react-testing-library/api/#render) function to produce actual DOM nodes. Then we define user data for reuse later on.
```

Author: - [x] Change "and" to "then"

**Note:** In this example we opt-in to destructuring the return value from our `render` function to give us our query functions. Optionally, you can utilize [screen](https://testing-library.com/docs/queries/about#screen) to access the query functions.
```js
// DOM Interaction and Queries
...
userEvent.type(getByLabelText(/username/i), user.username);
userEvent.type(getByLabelText(/password/i), user.password);
userEvent.click(getByText(/submit/i));
...
```

> Unit tests should be isolated, and your components should use a unique label and text for each element. With that in mind, we can use the query functions that we destructured from `render` to find the input elements in our component by their label text. `Testing Library` provides a utility called [userEvent](https://testing-library.com/docs/ecosystem-user-event/) that allows us to interact with the DOM nodes. We leverage the `type` interaction to enter our username and password into each respective field, then the `click` function to submit after querying with the submit button text.

Reviewer: ```suggestion
Unit tests should be isolated and your components should use a unique label and text for each test.

We can use the query functions that we destructured from `render` to find the input elements in our component by their label text. `Testing Library` provides a utility called [userEvent](https://testing-library.com/docs/ecosystem-user-event/) that allows us to interact with the DOM nodes. We leverage the `type` interaction to enter our username and password into each respective field, then the `click` function to submit after querying with the submit button text.
```

## Example unit tests - Enzyme
VSP provides helper utilities to make writing tests easier. The following example unit tests illustrate some of these helpers. You can find these unit tests in the `vets-website` repo in the [veteranInformation.unit.spec.jsx](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/vic-v2/tests/config/veteranInformation.unit.spec.jsx) file.
```js
import React from 'react';
import { expect } from 'chai';
import sinon from 'sinon';
import { mount } from 'enzyme';
import { DefinitionTester, fillData, selectRadio, fillDate } from '../../../../platform/testing/unit/schemaform-utils.jsx';
import formConfig from '../config/form.js';
describe('VIC veteran information', () => {
  const { schema, uiSchema } = formConfig.chapters.veteranInformation.pages.veteranInformation;
  it('should render', () => {
    const form = mount(
      <DefinitionTester
        definitions={formConfig.defaultDefinitions}
        schema={schema}
        data={{}}
        uiSchema={uiSchema}
        />
    );
    expect(form.find('input').length).to.equal(7);
    expect(form.find('select').length).to.equal(4);

>     form.unmount();

Reviewer: 🏆

describe('my-application', () => {
  describe('SimpleLoginForm', () => {
    it('calls onSubmit with the username and password when submit is clicked', () => {
        const handleSubmit = sinon.spy();
        const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
        const user = { username: 'user123', password: 'password123' };

        userEvent.type(getByLabelText(/username/i), user.username);
        userEvent.type(getByLabelText(/password/i), user.password);
        userEvent.click(getByText(/submit/i));

        expect(handleSubmit.callCount).to.equal(1); // alternatively `expect(handleSubmit.calledOnce).to.be.true()` works as well
        expect(handleSubmit.calledWith(user)).to.be.true(); // for more explicit testing we can use `calledWithExactly` in place of `calledWith`
    });
  });
});
```

We have successfully performed a "Happy Path" test for this component. Let's break down this test:

```js
// Setup
...
const handleSubmit = sinon.spy();
const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
const user = { username: 'user123', password: 'password123' };
...
```

> The setup for this test is self-explanatory. We create a mocked submit function and pass that to our form component, then use the [render](https://testing-library.com/docs/react-testing-library/api/#render) function to produce actual DOM nodes. Then we define user data for reuse later on.

Reviewer 3: ```suggestion
In setting up this test, we create a mocked submit function and pass that to our form component, then use the [render](https://testing-library.com/docs/react-testing-library/api/#render) function to produce actual DOM nodes. We then, define user data for reuse later on.
```

Author: - [x] Grammar/phrasing update

  );
};
export default SimpleLoginForm;
```

```js
// SimpleLoginForm.unit.spec.jsx
import React from 'react';
import { expect } from 'chai';
import sinon from 'sinon';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SimpleLoginForm from '../../components/SimpleLoginForm';
describe('my-application', () => {
  describe('SimpleLoginForm', () => {
    it('calls onSubmit with the username and password when submit is clicked', () => {
        const handleSubmit = sinon.spy();
        const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
        const user = { username: 'user123', password: 'password123' };

        userEvent.type(getByLabelText(/username/i), user.username);
        userEvent.type(getByLabelText(/password/i), user.password);
        userEvent.click(getByText(/submit/i));

        expect(handleSubmit.callCount).to.equal(1); // alternatively `expect(handleSubmit.calledOnce).to.be.true()` works as well
        expect(handleSubmit.calledWith(user)).to.be.true(); // for more explicit testing we can use `calledWithExactly` in place of `calledWith`
    });
  });
});
```

> We have successfully performed a "Happy Path" test for this component. Let's break down this test:

Reviewer 2: Nit.

```suggestion
We have written a "happy path" test for a `SimpleLoginForm` component. Let's break down this test.
```

Author: - [x] Phrasing fix

> 

Reviewer 2: To break this section for better readability, we could use subheaders describing what each part is doing. Something like "Setup", "DOM interactions and queries", and "Assertions".

```suggestion

#### Setup

```

Author: - [x] Formatting update

```js
// Setup
...
const handleSubmit = sinon.spy();
const { getByLabelText, getByText } = render(<SimpleLoginForm onSubmit={handleSubmit} />);
const user = { username: 'user123', password: 'password123' };
...
```

> In setting up this test, we create a mocked submit function and pass that to our form component, then use the [render](https://testing-library.com/docs/react-testing-library/api/#render) function to produce actual DOM nodes. We then, define user data for reuse later on.

Reviewer 2: Consider making bulleted lists for these types of paragraphs.

Reviewer 2: Something like this:

```suggestion
1. We mock the `handleSubmit` function.
2. We instantiate the `SimpleLoginForm` component, passing in the mocked `handleSubmit` as a prop.
3. We use the `render` function from RTL to produce actual DOM nodes.
4. We get the RTL query functions from destructuring the return value of `render`. Alternatively, you can utilize screen to access the query functions directly.
4. We define the user data for reuse later in the test.
```

Author: - [x] Formatting and phrasing

> **Note:** In this example we opt-in to destructuring the return value from our `render` function to give us our query functions. Optionally, you can utilize [screen](https://testing-library.com/docs/queries/about#screen) to access the query functions.

Reviewer 2: ```suggestion
**Note:** In this example we opted to destructure the return value from our `render` function to give us our query functions. Alternatively, you can utilize [screen](https://testing-library.com/docs/queries/about#screen) to access the query functions.
```

Author: - [x] Grammar fix

  false, // shouldResolve: false returns rejected promise. default is true
);
```
**Note:** This utility can be found in `platform/utilities/storage/localstorage.js`.
- `getLocalStorage()` - convenience accessor for local storage and implements a fallback. Useful for stubbing localstorage.

## Legacy tools
While ReactTestUtils, SkinDeep, and Enzyme are used in many of our tests, use React Testing Library for any updated or new tests.
## Libraries
Unit tests use the following libraries:
- [mocha.js](https://mochajs.org/): Test framework.
- [chai.js](https://www.chaijs.com/): BDD / TDD assertion library.
- [chai-as-promised](https://github.com/domenic/chai-as-promised): Extends Chai with assertions about promises.
- [sinon.js](https://sinonjs.org/): Standalone test spies, stubs and mocks.

> - [React Testing Library](https://testing-library.com/docs/react-testing-library/intro): React Testing Library is a branch of `Testing Library` that allows users to test their components through actual DOM nodes instead of rendered instances.

Reviewer: ```suggestion
- [React Testing Library](https://testing-library.com/docs/react-testing-library/intro): React Testing Library is a branch of `Testing Library` that allows users to test their React components through actual DOM nodes instead of rendered instances.
```

Author: - [x] Specify "React components" instead of just general "components"
